AIO basics blog post (#17298)
PostHog/posthog.com: contents/blog/what-is-ai-observability.mdx

---
title: 'What is AI observability, how it works, and which tools to use'
date: 2026-06-04
rootPage: /blog
sidebar: Blog
showTitle: true
hideAnchor: true
author:
    - natalia-amorim
featuredImage: >-

>   https://res.cloudinary.com/dmukukwp6/image/upload/posthog.com/contents/images/blog/hog_ql.png

Reviewer: We should use the robot construction/doctor image for this

Author: I don't know which one you're talking about @ivanagas is it this?

<img width="3844" height="2160" alt="doctor hogs blog" src="https://github.com/user-attachments/assets/44fb3a04-31c1-4dc8-b519-72fa0401975c" />

Author: <img width="2400" height="1260" alt="ai powered features" src="https://github.com/user-attachments/assets/5b36193d-ce8f-4f64-8213-a21b437ffb08" />

Author: Went with the second one cause I think it matches the vibe

Reviewer: yes

featuredImageType: full
category: General
tags:
    - ai-observability
seo:
    metaTitle: 'AI observability: what it is, how it works, and which tools to use'
    metaDescription: 'A comprehensive guide to AI observability: how it differs from traditional APM, what tools like PostHog actually do, and how to pick the right one for your stage.'
---

- The code works, the API calls succeed, but the outputs are off. 
- The model says the wrong thing. 
- The agent calls the wrong tool. 
- The bill is somehow 4x what you expected. 

None of this shows up in your normal monitoring, because your normal monitoring was built for software that does the same thing every time.

> That's the gap **AI observability** fills. It's the layer that watches what your LLMs are actually doing in production – what prompts you sent, what they returned, how long it took, how much it cost, and whether the output was any good.

Reviewer: ```suggestion
That's the gap **AI observability** fills. It's the layer that watches what your LLMs are actually doing in production – what prompts you sent, what they returned, how long it took, how much it cost, and whether the output was good.
```

> This guide walks through what AI observability is, the components it's built from, how the tools actually instrument your code, and which ones are worth using depending on your stack and stage.

Reviewer: 1. I think we need a first line before the list
2. We define AI observability is the next section so we don't need to define it again here
3. Tightening last paragraph

```suggestion
The problems with AI agents and apps often don't show up in normal monitoring tools:

- The code works, the API calls succeed, but the outputs are off. 
- The model says the wrong thing. 
- The agent calls the wrong tool. 
- The bill is 4x what you expected. 

That's because normal monitoring was built for software that does the same thing every time.

This gap is what **AI observability** fills. This guide walks through what AI observability is, the features and data it's built from, and which tools are worth using depending on your stack and stage.
```

## What is AI observability?

**AI observability** (often used interchangeably with **LLM observability** and **LLM analytics**) is the practice of monitoring the AI features of your application: prompts, responses, costs, latency, errors, and quality of outputs.

> It's a category that didn't really exist until the last couple of years, because the problems it solves didn't exist either. Traditional APM tools like [Datadog](/blog/best-datadog-alternatives) and New Relic were built to monitor software that behaves the same way every time.

Reviewer: ```suggestion
It's a category that barely existed two years ago, because the problems it solves didn't exist either. Traditional APM tools like [Datadog](/blog/best-datadog-alternatives) and New Relic were built to monitor software that behaves the same way every time.
```

AI features don't.

### How is it different from traditional monitoring

The short version:

| Traditional APM watches | AI observability watches |
|---|---|

> | Whether your API call succeeded | What the model actually said |

Reviewer: ```suggestion
| Whether API calls succeed | What the model said |
```

| Response times | Token usage and cost per call |
| Error rates | Quality of outputs (was it a hallucination?) |
| Service-level metrics | Agent-level metrics (which tools were called, in what order) |
| Did the code run? | Did the model give a useful answer? |

> Both layers are useful and you need both if you're running AI features in production, but they answer different questions.

Reviewer: I think this doesn't really say anything new, it's implied they're useful and answer different questions and need both
```suggestion
```

> The data model is different too. APM is built around requests, services, and traces of synchronous code paths. AI observability is built around *generations* (individual LLM calls), traces (the chain of generations and tool calls that make up an agent loop), and spans (the steps inside a trace).

Reviewer: I am assuming synchronous code paths only applies to traces? Traces usually do more than just that so this broadens that. 

```suggestion
The data model is different too. APM is built around requests, services, and traces across your application stack. AI observability is built around *generations* (individual LLM calls), traces (the chain of generations and tool calls that make up an agent loop), and spans (the steps inside a trace).
```

- **Data and ML teams** evaluating model performance at scale

If you have at least one LLM call in production, you need at least basic AI observability. 

We've written a more practical guide on [what to set up on day one if you're at the MVP stage](/blog/ai-observability-for-mvps), but the general rule is: tracing, cost tracking, and error capture, before anything else.

## How AI observability works

AI observability tools all work roughly the same way: they capture the data flowing through your LLM calls, store it in a queryable format, and give you tools to slice it, score it, and act on it. 

The differences are in what they capture, how they capture it, and what they let you do with it afterwards.

### The core building blocks

A good AI observability setup has six layers. Most tools cover the first three out of the box; the rest are either add-ons or features you build over time.

#### 1. Tracing

The foundation. Every LLM call (and every step inside an agent loop) gets captured as a [trace](/docs/ai-observability/traces): inputs, outputs, model, parameters, latency, token counts, tool calls, and any metadata you attach (user ID, feature, session).

> 

Reviewer: You want to add a screenshot of what a trace looks like here?

For simple apps with one LLM call, a trace is just one entry. For agents with multi-step workflows, a trace can include dozens of sub-spans showing the full chain of model calls, tool invocations, and decision points.

If you set up nothing else, set up tracing. Everything else builds on it.

#### 2. Cost and token tracking

> Once you have traces, cost tracking comes nearly for free – tools calculate cost per call from the token counts and model. The useful part is breaking it down: cost per user, per feature, per model, per session. That's how you spot when a power user is eating your margins or when a feature is calling the wrong model.

Reviewer: Would also be cool to show a screenshot of costs dashboard here too.

### The end-to-end workflow

A mature AI observability setup follows a loop:

1. **Capture** every call in production via tracing
2. **Monitor** cost, latency, and error rates with dashboards and alerts
3. **Evaluate** outputs with deterministic and LLM-as-judge evals, both offline (on test datasets during development) and online (on sampled production traces)
4. **Review** flagged traces manually – ones with low feedback scores, errors, support tickets, or outlier patterns
5. **Act** by turning what you learned into new evals, prompt changes, or product fixes
6. **Loop** – the new evals catch the next batch of regressions, you find new failure modes, and the cycle repeats

This is the workflow described in detail in [our beginner's guide to testing AI agents](/blog/testing-ai-agents). 

The point isn't to cover every possible input (you can't) but to make sure every bad interaction teaches your system something permanent.

> ## What are the best AI observability tools available

Reviewer: ```suggestion
## What are the best AI observability tools?
```

There are roughly half a dozen serious tools in this category as of 2026, plus several adjacent options (existing APM tools that added LLM features, ML monitoring tools that pivoted to AI observability). Here's the rundown.

### 1. PostHog

> [PostHog's AI observability](/ai-observability) is bundled with [product analytics](/product-analytics), [web analytics](/web-analytics), [session replay](/session-replay), [feature flags](/feature-flags), [experiments](/experiments), [surveys](/surveys), [error tracking](/error-tracking), [logs](/logs), and more – which means the same user model flows through all of them. 

Reviewer: ```suggestion
[PostHog's AI observability](/ai-observability) is bundled with [product analytics](/product-analytics), [web analytics](/web-analytics), [session replay](/session-replay), [feature flags](/feature-flags), [experiments](/experiments), [surveys](/surveys), [error tracking](/error-tracking), [logs](/logs), and more – which means the same user model connects through all of them. 
```

The free tier covers 100K LLM events/mo. EU hosting available. SDKs for OpenAI, Anthropic, LangChain, LlamaIndex, Vercel AI SDK, and others.

**Key features:**

- Tracing, cost tracking, latency, token usage, model breakdown
- [Evals](/docs/ai-evals/evaluations) with deterministic checks and LLM-as-a-judge
- [Prompt management](/docs/prompt-management/prompts) with versioning and fetch-at-runtime
- [Prompt experiments](/docs/prompt-management/prompt-experiments) (beta) for A/B testing prompts with built-in cost, latency, and eval pass rate metrics
- [MCP server](/docs/model-context-protocol) for querying observability data from Claude Code or Cursor
- Connected to [product analytics](/docs/product-analytics), [session replay](/docs/session-replay), and [error tracking](/docs/error-tracking) so you can debug AI failures with full user context

**Best for:** Teams that want everything in one place. Especially valuable for engineers focused on correlating model behavior with what users actually do and debugging AI features with user context.

### 2. Langfuse

> **Langfuse** is the open-source darling of the category. It's MIT-licensed, framework-agnostic, with depth across tracing, evaluations, and prompt management. Acquired by ClickHouse in early 2026.

Reviewer: ```suggestion
**Langfuse** is the open-source darling of the category. It's MIT-licensed, framework-agnostic, with depth across tracing, evaluations, and prompt management. It was acquired by ClickHouse in early 2026.
```

**Key features:**

- Tracing with detailed span visibility for agents
- Evals (both code-based and LLM-as-a-judge)
- Prompt management with versioning
- Self-hostable on your own infrastructure
- Strong integrations with LangChain, LlamaIndex, Haystack, and others

> **Best for:** Teams who want full control over their data through self-hosting, or who are building anything beyond raw API calls (agents, multi-step workflows, RAG pipelines). Self-host is free; the Core cloud plan starts at $29/mo.

Reviewer: Looks like they have a free cloud plan too: https://langfuse.com/pricing

### 3. LangSmith

**LangSmith** is built by the LangChain team and is the default observability tool if you're using LangChain or LangGraph. The agent debugging experience is strong if you live in that ecosystem.

**Key features:**

- Native LangChain and LangGraph integration
- Trace visualization for complex agent runs
- Prompt playground for iteration
- Evals with custom scorers
- Dataset management for offline evaluation

**Best for:** Teams already committed to LangChain. Pricing is seat-based ($39/user/mo on Plus) plus trace-based usage, which gets expensive quickly for larger teams.

### 4. Datadog LLM Observability

> [Datadog](/blog/best-datadog-alternatives) added LLM Observability as a product line in 2024, and it's evolved into one of the strongest enterprise option for teams already using Datadog for the rest of their stack.

Reviewer: ```suggestion
[Datadog](/blog/best-datadog-alternatives) added LLM Observability as a product line in 2024, and it's evolved into one of the strongest enterprise options for teams already using Datadog for the rest of their stack.
```

**Key features:**

- Tracing with correlation to backend services, infrastructure, and real user sessions
- Built-in evals and a sensitive data scanner for PII redaction
- Integrates with Datadog's broader observability suite (APM, logs, RUM)

> **Best for:** Teams already invested in the Datadog ecosystem who want LLM monitoring in the same pane as their backend services. The catch is pricing: the free tier covers 40K LLM spans/mo, but Datadog automatically activates LLM Observability charges the moment it detects LLM spans, so be careful if you're already on Datadog for other workloads. Hard to justify at the MVP stage.

Reviewer: This is a big complicated paragraph otherwise
```suggestion
**Best for:** Teams already invested in the Datadog ecosystem who want LLM monitoring in the same pane as their backend services. 

The catch is pricing: the free tier covers 40K LLM spans/mo, but Datadog automatically activates LLM Observability charges the moment it detects LLM spans, so be careful if you're already on Datadog for other workloads. Hard to justify at the MVP stage.
```

### 5. Portkey

**Portkey** is an AI gateway and observability platform that routes requests across 200+ LLM providers through one OpenAI-compatible endpoint. Setup is one base URL change, similar to the proxy pattern Helicone popularized.

**Key features:**

- Proxy-based logging with one-line setup
- Multi-provider routing with automatic fallback and load balancing
- Cost tracking and per-request logs
- Caching to reduce repeated calls
- Guardrails for content moderation and PII detection

> **Best for:** Teams who want cost tracking and basic logs with minimal setup, or who are juggling multiple LLM providers and want a unified routing layer. The trade-off, as with any proxy-based tool, is that you get less detail than SDK-based tracing for complex multi-step agents. Advanced features (some observability, security, governance) are gated to the cloud product. Free tier covers 10K requests/mo; paid plans start at $49/mo.

Reviewer: same
```suggestion
**Best for:** Teams who want cost tracking and basic logs with minimal setup, or who are juggling multiple LLM providers and want a unified routing layer. 

The trade-off, as with any proxy-based tool, is that you get less detail than SDK-based tracing for complex multi-step agents. Advanced features (some observability, security, governance) are gated to the cloud product. Free tier covers 10K requests/mo; paid plans start at $49/mo.
```

Reviewer: also, what does "some observability" mean?

| You want zero-setup cost logging across multiple providers | **Portkey** | Proxy-based, change one URL, you're done; supports 200+ providers |
| Your work is prompt-iteration-heavy | **Braintrust** | Eval-first, polished prompt workflow |

For a deeper dive into self-hostable options, see our roundup of [the best open source LLM observability tools](/blog/best-open-source-llm-observability-tools).

### Decision criteria worth caring about

If you're picking a tool, here's what actually matters:

- **Integration surface.** Does it work with OpenAI, Anthropic, the open-source model you're using, and the framework (if any) you've built on? Don't pick a tool you'll have to wrap yourself.
- **Cost model.** Per-event, per-user, per-span, seat-based – they all behave differently as you scale. Read the pricing page carefully.
- **Free tier reality.** "Free" means different things. Check what's actually included, what's gated, and how long data is retained.
- **Data ownership.** If you're handling sensitive data (PII, regulated industries), self-hosting or EU residency may matter more than features.

> - **Bundled vs unbundled.** Standalone tools are flexible but fragment user context. Bundled platforms ([PostHog](/) being the obvious example) keep everything in one user model, which makes correlation analyses possible.

Reviewer: ```suggestion
- **Bundled vs unbundled.** Standalone tools are flexible but fragment user context. Bundled platforms ([PostHog](/) being the obvious example) keep everything in one user model, which makes correlation analysis possible.
```

- **MCP support.** If you're shipping in 2026, your AI tools should let you query their data from Claude Code, Cursor, or other AI workflows. Tools that don't are already feeling dated.

## Why PostHog works well for AI observability

A quick pitch, since this is our blog:

> PostHog covers all six core layers (tracing, costs, errors, evals, feedback, prompts) in one tool, on one free tier (100K events/mo), with the same user model flowing through product analytics, session replay, feature flags, logs, surveys, and more. That means:

Reviewer: flowing makes it sound more temporary than it actually is I think
```suggestion
PostHog covers all six core layers (tracing, costs, errors, evals, feedback, prompts) in one tool, on one free tier (100K events/mo), with the same user model connected through product analytics, session replay, feature flags, logs, surveys, and more. That means:
```

</details>

<details>
<summary>How is AI observability different from regular APM?</summary>

**APM** tells you whether your code is running and how fast. It sees that you made an API call and whether the request was successful, but it can't see what the model said, whether it was good, or how much it cost. **AI observability** adds the missing layer: the content, the quality, and the economics.

You usually want both. PostHog covers both sides: [AI observability](/ai-observability) for what your models are doing, plus [error tracking](/error-tracking) and [logs](/logs) for the rest of the stack, so debugging an AI feature doesn't mean switching between three tools.

</details>

<details>
<summary>Does PostHog use its own AI observability internally?</summary>

Yes. The PostHog team uses AI observability, error tracking, and [logs](/blog/how-posthog-uses-logs) to debug our own features, including the PostHog Wizard and the agents that power our product analytics. 

</details>

<details>
<summary>Do I need AI observability if I just have one LLM call in my app?</summary>

> Yes. The setup is fast, and the free tiers will cover you well past a thousand users. [Here's what to set up if you're at the MVP stage](/blog/ai-observability-for-mvp-ai-apps).

Reviewer: ```suggestion
Yes. The setup is fast, and the free tiers will cover you well past a thousand users. [Here's what to set up if you're at the MVP stage](/blog/ai-observability-for-mvps).
```
